pkg/controller/nstemplatetier/nstemplatetier_controller.go

```diff
@@ -0,0 +1,226 @@
+		return reconcile.Result{}, err
+	} else if requeue {
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// ensureTemplateUpdateRequest ensures that all relared MasterUserRecords are up-to-date with the NSTemplateTier that changed.
+// If not, then it creates a TemplateUpdateRequest resource for the first MasterUserRecord not up-to-date with the tier, and
+// return `true, nil` so the controller will requeue the request to create subsequent TemplateUpdateRequest resources, until the
+// `MaxPoolSize` threashold is reached.
+func (r *ReconcileNSTemplateTier) ensureTemplateUpdateRequest(logger logr.Logger, tier *toolchainv1alpha1.NSTemplateTier) (bool, error) {
+	activeTemplateUpdateRequests, err := r.activeTemplateUpdateRequests(tier)
+	if err != nil {
+		return false, errs.Wrap(err, "unable to get active TemplateUpdateRequests")
+	}
+
+	if activeTemplateUpdateRequests < MaxPoolSize {
+		// create a TemplateUpdateRequest if active count < max pol size
+		// find a MasterUserRecord which is not already up-to-date
+		// and for which there is no TemplateUpdateRequest yet
+		// fetch by subsets of "MaxPoolSize + 1" size until a MasterUserRecord candidate is found
+		murs := toolchainv1alpha1.MasterUserRecordList{}
+		matchingLabels, err := murSelector(tier)
+		if err != nil {
+			return false, errs.Wrap(err, "unable to get MasterUserRecords to update")
+		}
+		err = r.client.List(context.Background(), &murs,
+			client.InNamespace(tier.Namespace),
+			client.Limit(MaxPoolSize+1),
+			matchingLabels,
+		)
+		if err != nil {
+			return false, errs.Wrap(err, "unable to get MasterUserRecords to update")
+		}
+		logger.Info("listed MasterUserRecords", "count", len(murs.Items), "selector", matchingLabels)
+		for _, mur := range murs.Items {
+			// check if there's already a TemplateUpdateRequest for this MasterUserRecord
+			templateUpdateRequest := toolchainv1alpha1.TemplateUpdateRequest{}
+			if err := r.client.Get(context.TODO(), types.NamespacedName{
+				Namespace: tier.Namespace,
+				Name:      mur.Name,
+			}, &templateUpdateRequest); err == nil {
+				logger.Info("MasterUserRecord already has an associated TemplateUpdateRequest", "name", mur.Name)
+				continue
+			} else if !errors.IsNotFound(err) {
+				return false, errs.Wrapf(err, "unable to get TemplateUpdateRequest for MasterUserRecord '%s'", mur.Name)
+			}
+			for _, account := range mur.Spec.UserAccounts {
+				accountTmpls := account.Spec.NSTemplateSet
+				logger.Info("checking MasterUserRecord account", "name", mur.Name, "tier", accountTmpls.TierName)
+				if accountTmpls.TierName != tier.Name {
+					// skip if the user account is not based on the current Tier
+					continue
+				}
+				// compare all TemplateRefs in the current MUR.UserAccount vs NSTemplateTier tier
+				if !reflect.DeepEqual(templateRefsFor(accountTmpls), templateRefsFor(tier.Spec)) {
```

Comment:
may I know why do you compare the NSTemplateSets in UserAccounts? if the MUR doesn't have the corresponding label, then it should be either updated or the label fixed

Reply:
indeed, this check was superfluous since we already selected MasterUserRecords with an outdated label for the updated tier. I removed it in d12fa5f